src/PlaneWaveBasis.jl

```diff
@@ -470,3 +480,65 @@ function PlaneWaveBasis(basis::PlaneWaveBasis; use_symmetry)
                                [[identity_symop()] for _ in 1:length(kcoords)];
                                fft_size=basis.fft_size)
 end
+
+
+"""
+Gather the distributed k-Point data on the master process and return
+it as a `PlaneWaveBasis`. On the other (non-master) processes `nothing` is returned.
+The returned object should not be used for computations and only to extract data
+for post-processing and serialisation to disk.
+"""
+function mpi_kgather(basis::PlaneWaveBasis)
+    # No need to allocate and setup a new basis object
+    mpi_nprocs(basis.comm_kpts) == 1 && return basis
+
+    # Gather k-Point info on master
+    kcoords = getproperty.(basis.kpoints, :coordinate)
+    kcoords = mpi_kgather(kcoords, basis)
+    ksymops = mpi_kgather(basis.ksymops, basis)
+
+    # Number of distinct k-Point coordinates is number of k-Points with spin 1
+    n_spinup_thisproc = count(isequal(1), getproperty.(basis.kpoints, :spin))
+    n_kcoords = mpi_sum(n_spinup_thisproc, basis.comm_kpts)
+
+    if isnothing(kcoords)  # i.e. master process
+        nothing
+    else
+        PlaneWaveBasis(basis.model,
+                       basis.Ecut,
+                       kcoords[1:n_kcoords],
+                       ksymops[1:n_kcoords],
+                       basis.symmetries;
+                       fft_size=basis.fft_size,
+                       kgrid=basis.kgrid,
+                       kshift=basis.kshift,
+                       variational=true,  # TODO hard-coded
+                       comm_kpts=MPI.COMM_SELF,
+                       instantiate_terms=false
+                      )
+    end
+end
+
+
+"""
+Gather the distributed data of a quantity depending on `k`-Points on the master process
+and return it. On the other (non-master) processes `nothing` is returned.
+"""
+function mpi_kgather(data::AbstractArray, basis::PlaneWaveBasis)
+    master = tag = 0
+    n_kpts = sum(length, basis.krange_allprocs)
+
+    if MPI.Comm_rank(basis.comm_kpts) == master
+        allk_data = similar(data, n_kpts)
+        allk_data[basis.krange_allprocs[1]] = data
+        for rank in 1:mpi_nprocs(basis.comm_kpts) - 1  # Note: MPI ranks are 0-based
```

Comment:
Don't you want to use MPI.gather here?

Reply:
(individual sends don't scale very well compared to a gather)

Reply:
I would love to, but it does not work, because you need an isbits type.

Reply:
Worth opening an issue at MPI, or converting to a canonical plain array? We probably need that at other places too.

Reply:
Does not work, because they are of different size for each k-Point.

Also I can contact the MPI folks, but I understand that they don't want to do it, because serialisation of arbitrary objects may lead to variable sizes on each process, that you need to communicate first before you can do a gatherv. So an MPI gather comes down to two gathers, which is perhaps surprising. Also it leads to copying quite a bit of data. So to me that is something that should be handled in the layer above MPI.jl.

Reply:
OK. Note there is a generalized MPI gather (gatherv) where you take into account different sizes.

Reply:
Yes but you still need isbits types. The problem is not just that you have different number of k-Points on each processor it's also that the \psi-Arrays (it's an array of arrays remember) themselves are of different size.